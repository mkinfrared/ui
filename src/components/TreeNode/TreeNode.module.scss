@import "../../styles/common";

.TreeNode {
  --duration: 250ms;
  --timing-function: ease-in-out;
  --depth: var(--group-depth, 0);
  --multiplier: calc(var(--depth) / 10);
  --node-padding: calc(var(--depth) * 2ch);
  --border-color: transparent;

  list-style: none;

  &:focus {
    outline: none;

    > .label {
      @include focus-visible;

      //outline-offset: -2px;
    }
  }

  .label {
    border-left: 0.4rem var(--border-color) solid;
    cursor: pointer;
    display: grid;
    gap: 0 1em;
    grid-template-columns: 2.4rem 1fr;
    padding-inline: var(--node-padding) 0;
    transition: background-color var(--duration) var(--timing-function);

    &:hover {
      background: hsl(0deg 100% 100% / 20%);
    }

    &.active {
      --border-color: var(--secondary);

      background: var(--purple-gradient);
    }

    &.expanded {
      .icon {
        transform: rotate(90deg);
      }
    }
  }

  .icon {
    fill: currentcolor;
    height: 100%;
    padding: 0.3em;
    transition: transform var(--duration);
    visibility: hidden;

    &.visible {
      visibility: visible;
    }
  }

  .group {
    --group-depth: calc(var(--depth) + 1);

    list-style: none;
    /* to overcome overflow-hidden issue
    *  add padding 4px and margin-block -4px (2px of outline on each side)
     */
    margin-block: -4px;
    overflow-y: hidden;
    padding: 4px;
    transition: height var(--duration) var(--timing-function);
  }
}
