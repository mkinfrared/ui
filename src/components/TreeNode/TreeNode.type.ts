import { ReactNode } from "react";

export type TreeNodeProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  children?: ReactNode;
  label: string;
  nodeId: string;
};

export type DoneCallback = () => void;
