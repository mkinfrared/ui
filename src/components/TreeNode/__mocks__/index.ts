export { default } from "./TreeNode.mock";
