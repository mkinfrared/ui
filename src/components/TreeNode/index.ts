export { default } from "./TreeNode";

export * from "./TreeNode.type";
