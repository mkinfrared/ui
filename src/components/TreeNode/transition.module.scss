.enter {
  height: 0;
}

.enterActive {
  height: var(--group-height);
}

.exit {
  height: var(--group-height);
}

.exitActive {
  height: 0;
}
