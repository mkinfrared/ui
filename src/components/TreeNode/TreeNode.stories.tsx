import { Meta, Story } from "@storybook/react";

import css from "./Story.module.scss";
import { TreeNode } from "./TreeNode";
import { TreeNodeProps } from "./TreeNode.type";

export default {
  title: "UI/TreeNode",
  component: TreeNode,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<TreeNodeProps> = (args) => (
  <div className={css.Story}>
    <TreeNode {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
