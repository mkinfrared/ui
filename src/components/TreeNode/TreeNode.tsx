/* eslint-disable jsx-a11y/role-has-required-aria-props */
import { KeyboardEventHandler, useContext } from "react";
import { CSSTransition } from "react-transition-group";

import FakeButton from "components/FakeButton";
import Heading from "components/Heading";
import TreeViewContext from "components/TreeView/TreeViewContext";
import { classNames } from "utils";

import { ReactComponent as Chevron } from "../../icons/chevron.svg";

import transition from "./transition.module.scss";
import css from "./TreeNode.module.scss";
import { DoneCallback, TreeNodeProps } from "./TreeNode.type";

const TreeNode = ({ children, className, label, nodeId }: TreeNodeProps) => {
  const { isActive, isExpanded, onNodeClick } = useContext(TreeViewContext);

  const handleClick = () => {
    onNodeClick?.(nodeId);
  };

  const handleAnimation = (node: HTMLElement, done: DoneCallback) => {
    node.addEventListener("transitionend", done, false);
  };

  const refCallback = (node: HTMLElement | null) => {
    if (!node) {
      return;
    }

    const { scrollHeight } = node;
    const { left } = node.getBoundingClientRect();

    node.style.setProperty("--group-height", `${scrollHeight}px`);

    node.style.setProperty("--group-left", `${left}px`);
  };

  const isExpandable = Array.isArray(children) ? !!children.length : !!children;

  const handleKeyDown: KeyboardEventHandler<HTMLLIElement> = (event) => {
    const { key, target, currentTarget } = event;

    if (key !== " ") {
      return;
    }

    if (target === currentTarget) {
      handleClick();
    }
  };

  const getAriaProps = () => {
    if (children) {
      return {
        "aria-expanded": !!isExpanded?.(nodeId),
        "aria-selected": !!isActive?.(nodeId),
      };
    }

    return { "aria-selected": !!isActive?.(nodeId) };
  };

  return (
    <li
      className={classNames(css.TreeNode, className)}
      data-nodeid={nodeId}
      data-testid="TreeNode"
      role="treeitem"
      onKeyDown={handleKeyDown}
      {...getAriaProps()}
    >
      <FakeButton className={css.label} onClick={handleClick} tabIndex={-1}>
        <Chevron
          className={classNames(css.icon, isExpandable && css.visible)}
        />
        <Heading variant="h4">{label}</Heading>
      </FakeButton>
      <CSSTransition
        in={!!isExpanded?.(nodeId)}
        addEndListener={handleAnimation}
        classNames={transition}
        unmountOnExit
      >
        <ul className={css.group} role="group" ref={refCallback} tabIndex={-1}>
          {children}
        </ul>
      </CSSTransition>
    </li>
  );
};

export { TreeNode };

export default TreeNode;
