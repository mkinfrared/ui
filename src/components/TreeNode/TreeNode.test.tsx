import { render } from "@testing-library/react";

import { TreeNode } from "./TreeNode";

describe("<TreeNode />", () => {
  const label = "Marklar";
  const nodeId = "foobar";
  const Component = <TreeNode nodeId={nodeId} label={label} />;

  it("should be defined", () => {
    expect(TreeNode).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("TreeNode");

    expect(element).toBeDefined();
  });
});
