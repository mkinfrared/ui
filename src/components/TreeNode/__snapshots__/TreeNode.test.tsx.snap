// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeNode /> should match the snapshot 1`] = `
<div>
  <li
    aria-expanded="false"
    aria-selected="false"
    class="TreeNode"
    data-nodeid="foobar"
    data-testid="TreeNode"
    role="treeitem"
  >
    <div
      class="FakeButton label"
      data-testid="FakeButton"
      role="button"
      tabindex="-1"
    >
      <svg
        class="icon visible"
      >
        chevron.svg
      </svg>
      <h4
        class="Heading"
        data-testid="Heading"
      >
        Marklar
      </h4>
    </div>
  </li>
</div>
`;
